src/wav2vec2/modeling.py

```diff
@@ -0,0 +1,311 @@
+        print("Total number of loaded variables:", len(model.variables))
+        return model
+
+    def _init(self, input_shape=None):
+        """Build Model weights using dummy inputs"""
+        # call this at the end only
+        if input_shape is None:
+            input_shape = (1, 2048)
+        dummy_input = tf.ones(input_shape, dtype=tf.float32)
+        return self(dummy_input)
+
+
+class Wav2Vec2Model(TFKerasModel):
+    def __init__(self, config: Wav2Vec2Config, name="wav2vec2"):
+        super().__init__(name=name)
+        if not isinstance(config, Wav2Vec2Config):
+            raise ValueError("`config` must be an instace of `Wave2Vec2Config`")
+
+        num_feature_extractor_layers = len(config.filter_sizes)
+
+        self.feature_extractor = [
+            FeatureExtractorLayer(
+                config, layer_id=i, name=f"feature_extractor/conv_layers/{i}"
+            )
+            for i in range(num_feature_extractor_layers)
+        ]
+        self.feature_projection = FeatureProjection(config, name="feature_projection")
+        self.encoder = Wav2Vec2Encoder(config, name="encoder")
+
+    def call(self, batch, training=False):
+        batch = tf.expand_dims(batch, axis=-1)
+        for feature_extractor_layer in self.feature_extractor:
+            batch = feature_extractor_layer(batch)
+        batch = self.feature_projection(batch, training=training)
+
+        # TODO: apply spec-augment to batch later (useful for training only)
+
+        batch = self.encoder(batch, training=training)
+        return batch
+
+    def freeze_feature_extractor(self):
+        for i in range(len(self.model.feature_extractor)):
+            self.model.feature_extractor[i].trainable = False
+
+
+class Wav2Vec2ForCTC(TFKerasModel):
+    """Wave2Vec2 model with a CTC head"""
+
+    def __init__(
+        self, config: Wav2Vec2Config, input_shape=(1, 2048), name="wav2vec-ctc"
+    ):
+        super().__init__(name=name)
+        if not isinstance(config, Wav2Vec2Config):
+            raise ValueError("`config` must be an instace of `Wave2Vec2Config`")
+        self.config = config
+
+        self.model = Wav2Vec2Model(config, name="wav2vec2")
+        self.dropout = tf.keras.layers.Dropout(config.dropout)
+        self.lm_head = tf.keras.layers.Dense(config.vocab_size, name="lm_head")
+
+        self._init(input_shape=input_shape)
+
+    def freeze_feature_extractor(self):
+        self.model.freeze_feature_extractor()
+        print("total number of trainable variables:", len(self.trainable_variables))
+
+    def call(self, batch, training=False):
+        # batch - (bsz, seqlen)
+        batch = self.model(batch, training=training)
+        batch = self.dropout(batch, training=training)
+        batch = self.lm_head(batch)
+
+        outputs = {"logits": batch}
+        if training:
+            # TODO: implement loss
+            loss = None
+            outputs.update({"loss": loss})
+
+        return outputs
```

Comment:
Since this seems to be the meat of the project would it make sense to modularize it a bit? Maybe consider grouping together the layers that have similar semantics and move them to a separate script? @MorganR what do you think?

Reply:
+1, I think at least splitting layers into a separate module from models would be helpful.